min and max pitch options

andrewharvey opened the pull request:
## Launch Checklist

 - [x] briefly describe the changes in this PR

add get/set min/maxPitch map options. This provides some groundwork for #3731, but doesn't change the maximum pitch you can set (still 60).

 - [ ] ~~include before/after visuals or gifs if this PR includes visual changes~~
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs
 - [ ] ~~post benchmark scores~~
 - [x] manually test the debug page

astojilj commented:
I'm not sure if it is overlapping with your patch: there's ongoing work on implementing [this](https://github.com/mapbox/mapbox-gl-native/pull/15230) in gl-js.
/cc @arindam1993

andrewharvey commented:
> I'm not sure if it is overlapping with your patch: there's ongoing work on implementing this in gl-js.

Oh I didn't see that. This PR doesn't do any of the fancy LOD features, it just adds get/set min/max zoom.

@arindam1993 if your work supersedes this, no worries, or if you want to build upon this that's fine. This could still be a nice to have feature even without the LOD (although it would become even more valuable with LOD and a higher maxzoom available).

Should I go ahead and add tests or leave as is until superseded by your work?

astojilj commented:
@andrewharvey 
#3731 is about supporting pitch larger than 60.
As mentioned, there's ongoing work on implementing larger value than 60 and that value will likely to get limited by the rendering engine itself - I see no need in providing additional control to user to increase the limit over the value supported by engine.
Is the use case here about specifying the subrange (of the range supported mapbox-gl-js) by the application developer - so that the application would allow pitch in e.g. range 0-30 degrees?

andrewharvey commented:
> As mentioned, there's ongoing work on implementing larger value than 60 and that value will likely get limited by the rendering engine itself

Sure, if possible it would be nice to have an update on #3731 about this, or just having that ticket assigned to someone at Mapbox to show it's currently being worked on.

> I see no need in providing additional control to user to increase the limit over the value supported by engine.

I agree, this PR specifically only lets you configure a min/max pitch within the bounds of what's supported. So right now 0-60.

> Is the use case here about specifying the subrange (of the range supported mapbox-gl-js) by the application developer - so that the application would allow pitch in e.g. range 0-30 degrees?

Yes that's exactly what this implements.

andrewharvey commented on src/geo/transform.js:
```diff
@@ -46,19 +46,24 @@ class Transform {
         this._maxZoom = maxZoom || 22;
 
+        this._minPitch = (minPitch === undefined || minPitch === null) ? 0 : minPitch;
+        this._maxPitch = (maxPitch === undefined || maxPitch === null) ? 60 : maxPitch;
```
I can't simply use `maxPitch || 60` because it will not accept `maxPitch: 0`.
I need to check both undefined and null because flow is picky.

andrewharvey commented on src/ui/map.js:
```diff
@@ -99,6 +101,11 @@ type MapOptions = {
+
+// the default values, but also the valid range
+const defaultMinPitch = 0;
+const defaultMaxPitch = 60;
```
Instead of defining both the default and limits, since the defaults are going to be the limits, I've used this constant for both the default and the limits.

andrewharvey commented on src/ui/map.js:
```diff
@@ -326,10 +338,22 @@ class Map extends Camera {
 
         if (options.minZoom != null && options.maxZoom != null && options.minZoom > options.maxZoom) {
-            throw new Error(`maxZoom must be greater than minZoom`);
+            throw new Error(`maxZoom must be greater than or equal to minZoom`);
```
I could split this out to a separate PR if needed as it is unrelated.

andrewharvey commented on test/unit/ui/map.test.js:
```diff
@@ -807,14 +807,103 @@ test('Map', (t) => {
         t.throws(() => {
             createMap(t, {minZoom:10, maxZoom:5});
-        }, new Error(`maxZoom must be greater than minZoom`));
+        }, new Error(`maxZoom must be greater than or equal to minZoom`));
```
part of the unrelated change earlier

andrewharvey commented:
I've fixed the failing tests. This does introduce a breaking change to the private Transform API, but since it's private and I've updated all internal use, is that an issue? To make it non-breaking I could change the order to just take on the new parameters to the end to avoid the breaking change.
